hpx/hpx_init_impl.hpp

```diff
@@ -355,78 +462,78 @@ namespace hpx
     {
         using hpx::program_options::options_description;
         options_description desc_commandline(
-            std::string("Usage: ") + app_name +  " [options]");
-
+            "Usage: " + app_name +  " [options]");
         util::function_nonser<int(hpx::program_options::variables_map& vm)>
             main_f;
-        std::vector<std::string> cfg;
         util::function_nonser<void()> const empty;
 
         HPX_ASSERT(argc != 0 && argv != nullptr);
 
-        return init(main_f, desc_commandline, argc, argv, cfg,
-            empty, empty, mode);
+        hpx::init_params iparams;
+        iparams.f = main_f;
+        iparams.desc_cmdline_ptr = std::make_shared<options_description>(
+            desc_commandline);
+        iparams.argc = argc;
+        iparams.argv = argv;
+        iparams.startup = empty;
+        iparams.shutdown = empty;
+        iparams.mode = mode;
+        return init(iparams);
     }
 
     inline int
     init(std::nullptr_t const& f,
         int argc, char** argv, hpx::runtime_mode mode)
     {
-        std::string app_name(HPX_APPLICATION_STRING);
-        return init(f, app_name, argc, argv, mode);
+        util::function_nonser<int(hpx::program_options::variables_map& vm)>
+            main_f;
+        util::function_nonser<void()> const empty;
+
+        HPX_ASSERT(argc != 0 && argv != nullptr);
+
+        hpx::init_params iparams;
+        iparams.f = main_f;
+        iparams.argc = argc;
+        iparams.argv = argv;
+        iparams.startup = empty;
+        iparams.shutdown = empty;
+        iparams.mode = mode;
+        return init(iparams);
     }
 
     inline int
     init(std::nullptr_t,
         int argc, char** argv, std::vector<std::string> const& cfg,
         hpx::runtime_mode mode)
     {
-        std::string app_name(HPX_APPLICATION_STRING);
-        using hpx::program_options::options_description;
-
-        options_description desc_commandline(
-            "Usage: " + app_name +  " [options]");
-
         util::function_nonser<int(hpx::program_options::variables_map& vm)>
             main_f;
 
         HPX_ASSERT(argc != 0 && argv != nullptr);
 
-        return init(main_f, desc_commandline, argc, argv, cfg,
-            startup_function_type(), shutdown_function_type(), mode);
+        hpx::init_params iparams;
+        iparams.f = main_f;
+        iparams.argc = argc;
+        iparams.argv = argv;
+        iparams.cfg = cfg;
+        iparams.mode = mode;
+        return init(iparams);
     }
 
     inline int
     init(std::nullptr_t, std::vector<std::string> const& cfg,
          hpx::runtime_mode mode)
     {
-        char* dummy_argv[2] = {
-            const_cast<char*>(HPX_APPLICATION_STRING), nullptr};
+        util::function_nonser<int(hpx::program_options::variables_map& vm)>
+            main_f;
 
-        return init(nullptr, 1, dummy_argv, cfg, mode);
+        hpx::init_params iparams;
+        iparams.f = main_f;
```

Comment:
Same here, I think you can just do 
```suggestion
        iparams.f = {};
```

Reply:
Why not have appropriate constructors defined for `hpx::init_params`?

Reply:
Actually we don't initialize the same parameters in each overload so there would be a lot of different constructors no ?
I agree that would simplify the code inside each overload but as a user I think I would prefer initialize it like that cause it is really fast to copy paste an `iparam.<sth>` assignment and change the 2 words (pretty fast in vim)
But if you want some I can totally make some ! :)